File: shipping-service/src/main/java/com/example/shipping_service/service/ShipmentService.java
package com.example.shipping_service.service;

import com.ecommerce.common.events.OrderPaidEvent;
import com.example.shipping_service.entity.Shipment;

public interface ShipmentService {

    Shipment createShipment(OrderPaidEvent event);

    Shipment getShipment(Long shipmentId);

    Shipment getShipmentByOrderId(Long orderId);

    Shipment markShipped(Long shipmentId,
                         String carrier,
                         String trackingNumber);

    Shipment markInTransit(Long shipmentId);

    Shipment markOutForDelivery(Long shipmentId);

    Shipment markDelivered(Long shipmentId);

    Shipment markFailed(Long shipmentId);

    Shipment cancelShipment(Long shipmentId);
}

File: ecommerce-frontend/src/services/shipmentService.js
import {
  apiGet,
  apiPost
} from "./api";

function getShipment(shipmentId) {
  return apiGet(
    `/shipments/${shipmentId}`
  );
}

function getShipmentByOrderId(orderId) {
  return apiGet(
    `/shipments/order/${orderId}`
  );
}

function trackShipmentByOrderId(orderId) {
  return apiGet(
    `/shipments/track/order/${orderId}`
  );
}

function markInTransit(shipmentId) {
  return apiPost(
    `/shipments/${shipmentId}/in-transit`
  );
}

function markOutForDelivery(shipmentId) {
  return apiPost(
    `/shipments/${shipmentId}/out-for-delivery`
  );
}

function markDelivered(shipmentId) {
  return apiPost(
    `/shipments/${shipmentId}/deliver`
  );
}

function markFailed(shipmentId) {
  return apiPost(
    `/shipments/${shipmentId}/fail`
  );
}

function cancelShipment(shipmentId) {
  return apiPost(
    `/shipments/${shipmentId}/cancel`
  );
}

export {
  getShipment,
  getShipmentByOrderId,
  trackShipmentByOrderId,
  markInTransit,
  markOutForDelivery,
  markDelivered,
  markFailed,
  cancelShipment
};